### ISSUE TITLE
BoundsError in eachoverlap()

### ISSUE THREAD

**Issue by yuifu** [BioJulia/XAM.jl#31]
## Expected Behavior

I tried to  count the number of BAM.Record objects overlapped with a specified genomic region by using `eachoverlap(bam_reader, chrom, leftpos:rightpos)`

```
julia> using GenomicFeatures

julia> using XAM

julia> using BioAlignments

julia> function test_bam_eachoverlap(path_bam, chrom, leftpos, rightpos)
	bam_reader = open(BAM.Reader, path_bam, index = path_bam * ".bai")
	count = 0
	for record in eachoverlap(bam_reader, chrom, leftpos:rightpos)
		count += 1
	end
	close(bam_reader)
	return count
end
```

This BAM file (`SRR7993829_1.100K.forward.bam`) has one record for JH584304.1:

```
$ samtools view SRR7993829_1.100K.forward.bam JH584304.1
SRR7993829.4835529.1    272     JH584304.1      51715   0       15M559N60M      *       0       0       CAAGCAGAACCTTTGGTCATGCCAGCGACAAGCCAACAGCCTGCGTAGGTGGCTGGAGCTAAGCCAGGAGGAGCG       JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFFFAA     AS:i:-5 ZS:i:-5 XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:74C0       YT:Z:UU XS:A:+  NH:i:2
```

So, the expected behavior is as follows:

```
julia> test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51000, 51200)
0

julia> test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51000, 51715)
1
```

## Current Behavior

I got an BoundsError:

```
julia> test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51000, 51200)
0

julia> test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51000, 51715)
ERROR: BoundsError: attempt to access 1-element Array{BGZFStreams.Block,1} at index [2]
Stacktrace:
 [1] getindex at ./array.jl:744 [inlined]
 [2] virtualoffset(::BGZFStreams.BGZFStream{IOStream}) at /Users/haruka/.julia/packages/BGZFStreams/qApsr/src/bgzfstream.jl:156
 [3] iterate(::XAM.BAM.OverlapIterator{IOStream}, ::XAM.BAM.OverlapIteratorState) at /Users/haruka/.julia/packages/XAM/9nk3g/src/bam/overlap.jl:65
 [4] test_bam_eachoverlap(::String, ::String, ::Int64, ::Int64) at ./REPL[83]:5
 [5] top-level scope at REPL[85]:1

julia> test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51715, 51716)
ERROR: BoundsError: attempt to access 1-element Array{BGZFStreams.Block,1} at index [2]
Stacktrace:
 [1] getindex at ./array.jl:744 [inlined]
 [2] virtualoffset(::BGZFStreams.BGZFStream{IOStream}) at /Users/haruka/.julia/packages/BGZFStreams/qApsr/src/bgzfstream.jl:156
 [3] iterate(::XAM.BAM.OverlapIterator{IOStream}, ::XAM.BAM.OverlapIteratorState) at /Users/haruka/.julia/packages/XAM/9nk3g/src/bam/overlap.jl:65
 [4] test_bam_eachoverlap(::String, ::String, ::Int64, ::Int64) at ./REPL[83]:5
 [5] top-level scope at REPL[86]:1
```

## Possible Solution / Implementation

I'm not sure how to fix the error, but I think that the error is relevant to the fact that the last record in this BAM file (`SRR7993829_1.100K.forward.bam`) is on JH584304.1:

```
$ samtools idxstats hisat2/SRR7993829_1.100K.forward.bam
chr1	195471971	1565	0
chr2	182113224	2116	0
chr3	160039680	1599	0
chr4	156508116	1789	0
chr5	151834684	1699	0
chr6	149736546	1486	0
chr7	145441459	1626	0
chr8	129401213	1393	0
chr9	124595110	1659	0
chr10	130694993	1230	0
chr11	122082543	2123	0
chr12	120129022	1580	0
chr13	120421639	1018	0
chr14	124902244	937	0
chr15	104043685	913	0
chr16	98207768	7585	0
chr17	94987271	2645	0
chr18	90702639	946	0
chr19	61431566	783	0
chrX	171031299	921	0
chrY	91744698	51	0
chrM	16299	730	0
GL456210.1	169725	0	0
GL456211.1	241735	0	0
GL456212.1	153618	0	0
GL456213.1	39340	0	0
GL456216.1	66673	0	0
GL456219.1	175968	0	0
GL456221.1	206961	0	0
GL456233.1	336933	0	0
GL456239.1	40056	0	0
GL456350.1	227966	2	0
GL456354.1	195993	0	0
GL456359.1	22974	0	0
GL456360.1	31704	0	0
GL456366.1	47073	0	0
GL456367.1	42057	0	0
GL456368.1	20208	0	0
GL456370.1	26764	0	0
GL456372.1	28664	0	0
GL456378.1	31602	0	0
GL456379.1	72385	0	0
GL456381.1	25871	0	0
GL456382.1	23158	0	0
GL456383.1	38659	0	0
GL456385.1	35240	0	0
GL456387.1	24685	0	0
GL456389.1	28772	0	0
GL456390.1	24668	0	0
GL456392.1	23629	0	0
GL456393.1	55711	0	0
GL456394.1	24323	0	0
GL456396.1	21240	0	0
JH584292.1	14945	0	0
JH584293.1	207968	1	0
JH584294.1	191905	6	0
JH584295.1	1976	0	0
JH584296.1	199368	0	0
JH584297.1	205776	0	0
JH584298.1	184189	1	0
JH584299.1	953012	0	0
JH584300.1	182347	0	0
JH584301.1	259875	0	0
JH584302.1	155838	0	0
JH584303.1	158099	0	0
JH584304.1	114452	1	0
ERCC-00002	1061	0	0
ERCC-00003	1023	0	0
ERCC-00004	523	0	0
ERCC-00009	984	0	0
ERCC-00012	994	0	0
ERCC-00013	808	0	0
ERCC-00014	1957	0	0
ERCC-00016	844	0	0
ERCC-00017	1136	0	0
ERCC-00019	644	0	0
ERCC-00022	751	0	0
ERCC-00024	536	0	0
ERCC-00025	1994	0	0
ERCC-00028	1130	0	0
ERCC-00031	1138	0	0
ERCC-00033	2022	0	0
ERCC-00034	1019	0	0
ERCC-00035	1130	0	0
ERCC-00039	740	0	0
ERCC-00040	744	0	0
ERCC-00041	1122	0	0
ERCC-00042	1023	0	0
ERCC-00043	1023	0	0
ERCC-00044	1156	0	0
ERCC-00046	522	0	0
ERCC-00048	992	0	0
ERCC-00051	274	0	0
ERCC-00053	1023	0	0
ERCC-00054	274	0	0
ERCC-00057	1021	0	0
ERCC-00058	1136	0	0
ERCC-00059	525	0	0
ERCC-00060	523	0	0
ERCC-00061	1136	0	0
ERCC-00062	1023	0	0
ERCC-00067	644	0	0
ERCC-00069	1137	0	0
ERCC-00071	642	0	0
ERCC-00073	603	0	0
ERCC-00074	522	0	0
ERCC-00075	1023	0	0
ERCC-00076	642	0	0
ERCC-00077	273	0	0
ERCC-00078	993	0	0
ERCC-00079	644	0	0
ERCC-00081	534	0	0
ERCC-00083	1022	0	0
ERCC-00084	994	0	0
ERCC-00085	844	0	0
ERCC-00086	1020	0	0
ERCC-00092	1124	0	0
ERCC-00095	521	0	0
ERCC-00096	1107	0	0
ERCC-00097	523	0	0
ERCC-00098	1143	0	0
ERCC-00099	1350	0	0
ERCC-00104	2022	0	0
ERCC-00108	1022	0	0
ERCC-00109	536	0	0
ERCC-00111	994	0	0
ERCC-00112	1136	0	0
ERCC-00113	840	0	0
ERCC-00116	1991	0	0
ERCC-00117	1136	0	0
ERCC-00120	536	0	0
ERCC-00123	1022	0	0
ERCC-00126	1118	0	0
ERCC-00130	1059	0	0
ERCC-00131	771	0	0
ERCC-00134	274	0	0
ERCC-00136	1033	0	0
ERCC-00137	537	0	0
ERCC-00138	1024	0	0
ERCC-00142	493	0	0
ERCC-00143	784	0	0
ERCC-00144	538	0	0
ERCC-00145	1042	0	0
ERCC-00147	1023	0	0
ERCC-00148	494	0	0
ERCC-00150	743	0	0
ERCC-00154	537	0	0
ERCC-00156	494	0	0
ERCC-00157	1019	0	0
ERCC-00158	1027	0	0
ERCC-00160	743	0	0
ERCC-00162	523	0	0
ERCC-00163	543	0	0
ERCC-00164	1022	0	0
ERCC-00165	872	0	0
ERCC-00168	1024	0	0
ERCC-00170	1023	0	0
ERCC-00171	505	0	0
*	0	0	0
```

## Steps to Reproduce (for bugs)

1. You can donwload the BAM file from https://www.dropbox.com/sh/8kenlaq25wenmvp/AAC622OEyG4RsuvriUNInfLMa?dl=0
2. The code is below:

```
using GenomicFeatures
using XAM
using BioAlignments

function test_bam_eachoverlap(path_bam, chrom, leftpos, rightpos)
	bam_reader = open(BAM.Reader, path_bam, index = path_bam * ".bai")
	count = 0
	for record in eachoverlap(bam_reader, chrom, leftpos:rightpos)
		count += 1
	end
	close(bam_reader)
	return count
end

test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51000, 51715)
```

## Your Environment

- Package Version used: BioAlignments v2.0.0, GenomicFeatures v2.0.3, XAM v0.2.6
- Julia Version used: 1.3.0 and 1.5.1
- Operating System and version (desktop or mobile): macOS 10.13.6

**Comment by zhixingfeng** [comment]
Hi Yuifu,
Got the same error on a different dataset. Have you got any feedback?

**Comment by yuifu** [comment]
@zhixingfeng Unfortunately, I've received no feedback so far.

**Comment by kescobo** [comment]
Unfortunately, this package has one primary author, and that person hasn't been around much. If you have any interest in working on this, you're more than welcome to. I don't work with these files, so I'm afraid I can't be of much help.

**Comment by CiaranOMara** [comment]
@yuifu and @zhixingfeng, does the [hotfix/issue-31](https://github.com/BioJulia/XAM.jl/tree/hotfix/issue-31) branch work with your datasets?

```julia
julia> add XAM#hotfix/issue-31
```

**Comment by zhixingfeng** [comment]
@CiaranOMara Thank you for your prompt feedback. The hotfix/issue-31 branch failed on my dataset as well as the test dataset of the updated XAM from branch hotfix/issue-31 with the same error message "BoundsError: attempt to access 1-element Array{BGZFStreams.Block,1} at index [2]". I tested on CentOS 7.6 on a AMD64 machine with Julia 1.5.3. Please find the error messages below

(v1.5) pkg> add XAM#hotfix/issue-31
   Updating git-repo `https://github.com/BioJulia/XAM.jl.git`
  Resolving package versions...
Updating `/sc/arion/projects/LymeMIND2/.julia/packages/environments/v1.5/Project.toml`
  [d759349c] + XAM v0.2.6 `https://github.com/BioJulia/XAM.jl.git#hotfix/issue-31`
Updating `/sc/arion/projects/LymeMIND2/.julia/packages/environments/v1.5/Manifest.toml`
  [67c07d97] + Automa v0.8.0
  [28d598bf] + BGZFStreams v0.3.0
  [00701ae9] + BioAlignments v2.0.0
  [37cfa864] + BioCore v2.0.5
  [47718e42] + BioGenerics v0.1.0
  [7e6ae17a] + BioSequences v2.0.5
  [3c28c6f8] + BioSymbols v4.0.4
  [e1450e63] + BufferedStreams v1.0.0
  [944b1d66] + CodecZlib v0.7.0
  [861a8166] + Combinatorics v1.0.2
  [899a7d2d] + GenomicFeatures v2.0.4
  [1cb3b9ac] + IndexableBitVectors v1.0.0
  [4ffb77ac] + Indexes v0.1.1
  [524e6230] + IntervalTrees v1.0.0
  [860ef19b] + StableRNGs v0.1.2
  [3bb67fe8] + TranscodingStreams v0.9.5
  [7200193e] + Twiddle v1.1.2
  [d759349c] + XAM v0.2.6 `https://github.com/BioJulia/XAM.jl.git#hotfix/issue-31`
  [ddb6d928] + YAML v0.4.6
  [9abbd945] + Profile


(v1.5) pkg> test XAM
    Testing XAM
Status `/tmp/jl_8Opd5y/Project.toml`
  [67c07d97] Automa v0.8.0
  [28d598bf] BGZFStreams v0.3.0
  [00701ae9] BioAlignments v2.0.0
  [47718e42] BioGenerics v0.1.0
  [7e6ae17a] BioSequences v2.0.5
  [3372ea36] FormatSpecimens v1.0.1
  [899a7d2d] GenomicFeatures v2.0.4
  [4ffb77ac] Indexes v0.1.1
  [3bb67fe8] TranscodingStreams v0.9.5
  [d759349c] XAM v0.2.6 `https://github.com/BioJulia/XAM.jl.git#hotfix/issue-31`
  [de0858da] Printf
  [8dfed614] Test
Status `/tmp/jl_8Opd5y/Manifest.toml`
  [56f22d72] Artifacts v1.3.0
  [67c07d97] Automa v0.8.0
  [28d598bf] BGZFStreams v0.3.0
  [00701ae9] BioAlignments v2.0.0
  [37cfa864] BioCore v2.0.5
  [47718e42] BioGenerics v0.1.0
  [7e6ae17a] BioSequences v2.0.5
  [3c28c6f8] BioSymbols v4.0.4
  [e1450e63] BufferedStreams v1.0.0
  [944b1d66] CodecZlib v0.7.0
  [861a8166] Combinatorics v1.0.2
  [34da2185] Compat v3.25.0
  [864edb3b] DataStructures v0.18.9
  [3372ea36] FormatSpecimens v1.0.1
  [899a7d2d] GenomicFeatures v2.0.4
  [1cb3b9ac] IndexableBitVectors v1.0.0
  [4ffb77ac] Indexes v0.1.1
  [524e6230] IntervalTrees v1.0.0
  [692b3bcd] JLLWrappers v1.2.0
  [bac558e1] OrderedCollections v1.4.0
  [860ef19b] StableRNGs v0.1.2
  [3bb67fe8] TranscodingStreams v0.9.5
  [7200193e] Twiddle v1.1.2
  [d759349c] XAM v0.2.6 `https://github.com/BioJulia/XAM.jl.git#hotfix/issue-31`
  [ddb6d928] YAML v0.4.6
  [83775a58] Zlib_jll v1.2.11+18
  [2a0f44e3] Base64
  [ade2ca70] Dates
  [8bb1440f] DelimitedFiles
  [8ba89e20] Distributed
  [b77e0a4c] InteractiveUtils
  [76f85450] LibGit2
  [8f399da3] Libdl
  [37e2e46d] LinearAlgebra
  [56ddb016] Logging
  [d6f4376e] Markdown
  [a63ad114] Mmap
  [44cfe95a] Pkg
  [de0858da] Printf
  [9abbd945] Profile
  [3fa0cd96] REPL
  [9a3f8284] Random
  [ea8e919c] SHA
  [9e88b42a] Serialization
  [1a1011a3] SharedArrays
  [6462fe0b] Sockets
  [2f01184e] SparseArrays
  [10745b16] Statistics
  [8dfed614] Test
  [cf7118a7] UUIDs
  [4ec0a83e] Unicode
Test Summary: | Pass  Total
SAM           |  122    122
BAM: Error During Test at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/test/test_bam.jl:286
  Test threw exception
  Expression: count((r->begin
                true
            end), eachoverlap(reader, "JH584304.1", 51000:51715)) == 1
  BoundsError: attempt to access 1-element Array{BGZFStreams.Block,1} at index [2]
  Stacktrace:
   [1] getindex at ./array.jl:809 [inlined]
   [2] virtualoffset(::BGZFStream{IOStream}) at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/BGZFStreams/qApsr/src/bgzfstream.jl:156
   [3] iterate(::XAM.BAM.OverlapIterator{IOStream}, ::XAM.BAM.OverlapIteratorState) at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/src/bam/overlap.jl:65
   [4] _simple_count at ./reduce.jl:869 [inlined]
   [5] count at ./reduce.jl:864 [inlined]
   [6] (::var"#9#21")(::XAM.BAM.Reader{IOStream}) at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/test/test_bam.jl:286
   [7] open(::var"#9#21", ::Type{XAM.BAM.Reader}, ::String; kwargs::Base.Iterators.Pairs{Symbol,String,Tuple{Symbol},NamedTuple{(:index,),Tuple{String}}}) at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/BioGenerics/cCuGr/src/IO.jl:48
   [8] top-level scope at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/test/test_bam.jl:283
   [9] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [10] top-level scope at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/test/test_bam.jl:2
  
Test Summary:              | Pass  Error  Total
BAM                        |  173      1    174
  AuxData                  |   11            11
  Record                   |    3             3
  Reader                   |   52            52
  Read long CIGARs         |    3             3
  Round trip               |   28            28
  In-Place-Reading Pattern |    7             7
  Random access            |   68            68
ERROR: LoadError: LoadError: Some tests did not pass: 173 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/test/test_bam.jl:1
in expression starting at /hpc/users/fengz03/LymeMind2/.julia/packages/packages/XAM/hu7aW/test/runtests.jl:27
ERROR: Package XAM errored during testing

**Comment by zhixingfeng** [comment]
@CiaranOMara. As @yuifu pointed out, the error might be caused by inquiring the last record in the bam file. Tried to quit eachoverlap loop by checking eof(), it worked to avoid the BoundError bug on my data and @yuifu 's data. See the code below.


using GenomicFeatures
using XAM
using BioAlignments

function test_bam_eachoverlap(path_bam, chrom, leftpos, rightpos)
	bam_reader = open(BAM.Reader, path_bam, index = path_bam * ".bai")
	count = 0
	for record in eachoverlap(bam_reader, chrom, leftpos:rightpos)
		count += 1
        **eof(bam_reader) && break**
	end
	close(bam_reader)
	return count
end

test_bam_eachoverlap("SRR7993829_1.100K.forward.bam", "JH584304.1", 51000, 51715)

**Comment by CiaranOMara** [comment]
With this dataset, the `VirtualOffset` appears to jump to a position before the starting offset (`chunk.start`). This means that `eof` could be reached and then the next chunk could move the offset to an earlier position. I don't think there is a requirement for chunks to be ordered, is there?

**Comment by CiaranOMara** [comment]
Update: I had another chance to continue bisecting the code to track down the error. It seems unlikely that seek does not work in BGZFStreams or @jakobnissen's new CodecBGZF. My current thinking is that the error lies in the offsets produced by the [overlapchunks](https://github.com/BioJulia/Indexes.jl/blob/e62f6c8c31c6a25079e8dfa71b8add53d0750b82/src/bgzfindex.jl#L40-72) method in Indexes.jl. My reasoning is that each offset should coincide with the beginning of a chunk, from which at least one byte should be available for reading.

**Comment by Marlin-Na** [comment]
Hi @CiaranOMara, I encountered this issue as well and I put together a BAM file that has two reads with the same position for reproducing the issue. I find that the `eof` check fix (e.g. with `feature/CodecBGZF-issue-31` branch) will actually skip one of the reads. See https://gist.github.com/Marlin-Na/9cb7767dfc87b007500a6fff4a8f5b9a

**Comment by Marlin-Na** [comment]
I think my issue was that if a read is unmapped, its `rightposition(record)` will become `position(record) - 1`, which is smaller than the leftposition. I think the check here https://github.com/BioJulia/XAM.jl/blob/develop/src/bam/overlap.jl#L87 should account for this special case. Probably change it to the following? It worked for me.
```
    if rid < interval[1] || (rid == interval[1] && rightposition(record) < first(interval[2]) &&  position(record) < first(interval[2]))
```
and
```
    if rid > interval[1] || (rid == interval[1] && position(record) > last(interval[2]) && rightposition(record) > last(interval[2]))
```

**Comment by lysogeny** [comment]
I believe that this is actually an issue of `BGZFStreams.jl` which I have tried to understand here: https://github.com/BioJulia/BGZFStreams.jl/issues/22. 
As the project's maintainers don't really seem to be responding, I am using my own fork: https://github.com/lysogeny/BGZFStreams.jl

To those of you with this problem does the proposed fix from there solve your issues?

**Comment by CiaranOMara** [comment]
@yuifu and @zhixingfeng, I had another chance to have a look at this issue, does the hotfix/issue-31-bytesavailable branch work with your datasets?

julia> add XAM#hotfix/issue-31-bytesavailable

**Comment by CiaranOMara** [comment]
Update:
The [eof check](https://github.com/BioJulia/XAM.jl/tree/hotfix/issue-31-eof) appears thread-dependent. The check appears to work for up to 2 threads.

**Comment by CiaranOMara** [comment]
@yuifu, @lysogeny, @zhixingfeng, and @BioJulia/admin, I believe I've tracked down the cause of this issue.

If you could, please review the current [hotfix/issue-31](https://github.com/BioJulia/XAM.jl/tree/hotfix/issue-31) branch and provide feedback.

```julia
julia> add XAM#hotfix/issue-31
```

As far as I'm aware, the fix addresses all of the relevant symptoms mentioned in this thread.

**Comment by CiaranOMara** [comment]
Should be fixed with #50.
